napari/utils/key_bindings.py

```diff
@@ -182,6 +187,257 @@ def normalize_key_combo(key_combo):
 
 
+def call_with_context(function, context):
+    """
+    call function `function` with the corresponding value taken from context
+
+    This is use in the action manager to pass the rights instances to the actions,
+    without having the need for them to take a **kwarg, and is mostly needed when
+    triggering actions via buttons, or to record.
+
+    If we went a declarative way to trigger action we cannot refer to instances
+    or objects that must be passed to the action, or at least this is
+    problematic.
+
+    We circumvent this by having a context (dictionary of str:instance) in
+    the action manager, and anything can tell the action manager "this is the
+    current instance a key". When an action is triggered; we inspect the
+    signature look at which instances it may need and pass this as parameters.
+    """
+    from inspect import signature
+
+    context_keys = [
+        k
+        for k, v in signature(function).parameters.items()
+        if v.kind not in (v.VAR_POSITIONAL, v.VAR_KEYWORD)
+    ]
+    ctx = {k: v for k, v in context.items() if k in context_keys}
+    return function(**ctx)
+
+
+@dataclass
+class Action:
+    command: Callable
+    description: str
+    keymappable: Any
+
+
+class ActionManager:
+    """
+    Manage the bindings between buttons; shortcuts, callbacks gui elements...
+
+    The action manager is aware of the various buttons, keybindings and other
+    elements that may trigger an action and is able to synchronise all of those.
+    Thus when a shortcut is bound; this should be capable of updating the
+    buttons tooltip menus etc to show the shortcuts, descriptions...
+
+    In most cases this should also allow to bind non existing shortcuts,
+    actions, buttons, in which case they will be bound only once the actions are
+    registered.
+
+    As the action manager also knows about all the UI elements that can trigger
+    an action, it can also be used to flash the buttons when the action is
+    triggered separately.
+
+    For actions that need access to a global element (a viewer, a plugin, ... ),
+    you want to give this item a unique name, and add it to the action manager
+    `context` object.
+
+    >>> action_manager.context['number'] = 1
+    ... action_manager.context['qtv'] = viewer.qt_viewer
+
+    >>> def callback(qtv, number):
+    ...     qtv.dims[number] +=1
+
+    >>> action_manager.register_action('bump one', callback,
+    ...     'Add one to dims',
+    ...     None)
+
+    The callback signature is going to be inspected and required globals passed
+    in.
+    """
+
+    _actions: Dict[str, Action]
+
+    def __init__(self):
+        # map associating a name/id with a Comm
+        self._actions = {}
+        self._buttons = defaultdict(lambda: [])
+        self._qactions = defaultdict(lambda: [])
+        self._shortcuts = {}
+        self._orphans = {}
+        self.context = {}
+
+    def register_action(self, name, command, description, keymappable):
```

Comment:
@Carreau some questions:
- Do we need a separate `id_` and `name`, where `name` is localizable and `id_` is the unique action identifier?
- This `name` is what is shown in `menus`, but under some cases we might want to have a toolbutton that has a name, that is different from name. Qt provides an `icon_text` for tool buttons that include a text. 
- How do we differentiate `toggles`  from `triggers`? `command` assumes we are only doing triggering.
- Where/how do we handle icons?
- How do we handle context (Qt.ApplicationShortcut/Qt.WidgetShortcut etc...)